Framework: Extract Layout out of the main bundle

alexsanford opened the pull request:
The `Layout` component is only used for logged-in users. It also has many dependencies which contribute significantly to the size of the main JavaScript bundle. By extracting it and loading it asynchronously, we reduce the size of the main bundle, and avoid loading the unneeded Layout for logged-out users.

Before extracting `Layout`:

<img width="210" alt="2017-06-10_1312 before" src="https://user-images.githubusercontent.com/842193/27004363-97f4a932-4dde-11e7-80b0-672bdce5dac3.png">

After extracting `Layout`:

<img width="210" alt="2017-06-10_1312 after-build" src="https://user-images.githubusercontent.com/842193/27004365-a25a1c5e-4dde-11e7-8705-b4123b68489d.png">

<img width="270" alt="2017-06-10_1312 after-layout" src="https://user-images.githubusercontent.com/842193/27004366-a405e6be-4dde-11e7-898b-982554270b92.png">

_**Update - Here are the full bundle reports**_

_Note: the numbers are different from above because I rebased after the initial reports were made_

- [Before extracting `Layout`](https://alexsanford.gitlab.io/calypso-performance/report-before-extract-layout-0316d6b7.html)

- [After extracting `Layout`](https://alexsanford.gitlab.io/calypso-performance/report-after-extract-layout-28922e7c.html)

Things to discuss/investigate:

- Is the async load going to have a negative effect on logged-in users? On one hand, the smaller `build.js` bundle should load more quickly, which will give an indication of progress sooner. On the other hand, the `Layout` itself might be rendered later because of the async load. If this is a problem, we could add a `script` tag in the HTML to load `Layout` right away for logged-in users.

- I changed a few places where `Layout` was rendered in order to render it asynchronously. It looks like everything still works (logging out, logging in, navigating, tests all pass), but I would like some more eyes on that code to be sure.

- Theoretically, we could also extract the `LayoutLoggedOut` component very easily. I didn't do so for a couple of reasons:

  1. It contributes much less size to the bundle. Using the [Bonsai](https://pinterest.github.io/bonsai/analyze/) analysis tool, `Layout` has a "Weighted Size" of `2.29 MB`, whereas `LayoutLoggedOut` has a "Weighted Size" of only `10.98 KB`.

  2. In the interest of keeping logged-out pages fast, I felt that loading `LayoutLoggedOut` async might have too much overhead. It is probably faster if we leave it in `build.js`.

alexsanford commented:
> Thanks for working on this @alexsanford. I am worried about the general usefulness of the change. Our main goals is to render something, like a decent the shell as quickly as possible.

Granted, it's definitely not a sweeping performance improvement, although I think it's a pretty good improvement for a first-time visitor. First-time visitors are generally going to be logged out, so they don't need to download this fairly large chunk of `Layout` code until they signup or login. @gziolo what do you think?

As for rendering quickly for repeat visitors, I think one thing we could do is split the bundle into smaller chunks to minimize the impact of code changes on the browser cache. See #14922

alexsanford commented on client/boot/project/wordpress-com.js:
```diff
@@ -32,19 +32,21 @@ const config = require( 'config' ),
 
-function renderLayout( reduxStore ) {
-	const Layout = require( 'controller' ).ReduxWrappedLayout;
+function renderLayoutPromise( reduxStore ) {
```
Yeah, I see what you mean. Changed it back to `renderLayout` for now. I think I was trying to make it clear that it was async, but I don't think that's necessary.

nb replied:
Maybe including “async” in the name will make it clearer?

alexsanford replied:
Done. `renderLayoutAsync` it is! 😄

nb replied:
👍

alexsanford commented on client/boot/project/wordpress-com.js:
```diff
@@ -32,19 +32,21 @@ const config = require( 'config' ),
-function renderLayout( reduxStore ) {
-	const Layout = require( 'controller' ).ReduxWrappedLayout;
+function renderLayoutPromise( reduxStore ) {
+	return require( 'controller' ).getReduxWrappedLayout( reduxStore ).then(
```
I'm not sure. I looked into how `require.ensure` handles errors, and it looks like we actually aren't doing any error handling for any of our `require.ensure` calls across the app. Are there other failure cases we need to look at here?

nb replied:
Can’t the request to the async chunk fail? What does webpack offer us?

alexsanford replied:
@nb Yeah, it could definitely fail. I've looked further into this, and here's what I have.

As far as I can tell, webpack 1 doesn't give us any error handling. There's a bit of a rabbit trail of issues and PR's around this, but nothing seems to have made it into webpack 1. Here are some links if you're interested:

- https://github.com/webpack/webpack/issues/758
- https://github.com/webpack/webpack/pull/4069
- https://github.com/webpack/webpack/pull/692
- https://github.com/webpack/webpack/pull/763
- https://github.com/webpack/webpack/pull/785

There are some third party plugins that look like they would provide error handling, but I don't know if we need/want to go there:

- https://github.com/NekR/async-module-loader
- https://github.com/richardscarrott/require-error-handler-webpack-plugin

Webpack 2 does have support for error handling of async loads through Promises.

Also, as far as I can tell, none of our current async loading has any error handling. In particular, I looked [here](https://github.com/Automattic/wp-calypso/blob/0ca65f4bddbefb8375f33fe7cbd5ddf736604597/server/bundler/babel/babel-plugin-transform-wpcalypso-async/test/index.js#L35) to see how `AsyncLoad` works, and [here](https://github.com/Automattic/wp-calypso/blob/0ca65f4bddbefb8375f33fe7cbd5ddf736604597/server/bundler/babel/babel-plugin-transform-wpcalypso-async/test/index.js#L35) to see how `preload` works, and neither appear to have error handling.

alexsanford replied:
@nb Also, here's a PR by @blowery on the subject: #12005

nb replied:
Thanks for the research, @alexsanford, it’s a rabbit hole, indeed. I guess it’s one more item to be punted for until webpack2 support gets in.

alexsanford commented on client/boot/project/wordpress-com.js:
```diff
@@ -85,7 +87,11 @@ export function setupMiddlewares( currentUser, reduxStore ) {
 	// Isomorphic sections will take care of rendering their Layout last themselves.
 	if ( ! document.getElementById( 'primary' ) ) {
-		renderLayout( reduxStore );
+		page( '*', function( context, next ) {
```
Hmm... probably we don't. I need to think a little more about this part. The problem is that it was rendering the Layout synchronously, and I needed to make it async.

My understanding is that this code should render the Layout only if the server-side hasn't already done so. Maybe I should have the check for the `primary` element *inside* of the middleware, so it only renders if it needs to.

@ockham do have some insight here?

gziolo replied:
You need to perform rendering on the server and then repeat on the client when SSR is enabled. Technically React discovers that DOM hasn't changed and skips repaint and reflow phase. I think we shouldn't introduce this change.

alexsanford replied:
I updated this. Would love your thoughts on the new implementation.

The problem is that on multi-tree pages that are not server-rendered, we need to render `Layout` before the section tries to render itself. Otherwise, when it tries to do so, the `#primary` and `#secondary` elements are not there for it to render itself into.

This wasn't a problem before when `renderLayout` was synchronous; we just rendered it directly inside of `setupMiddlewares`. But now that we're rendering the layout asynchronously, I think we need to do it in middleware to defer the call of `next()` until after the Layout is rendered.

The new middleware does so conditionally now, so it won't re-render every time.

cc @nb @gziolo

nb replied:
Thinking through all the different cases sounds great, but I am still a bit confused – why do we need to use `page( '*'…` at all?

alexsanford commented on client/boot/project/wordpress-com.js:
```diff
@@ -247,12 +253,16 @@ export function setupMiddlewares( currentUser, reduxStore ) {
 			debug( 'Re-rendering multi-tree layout' );
 			ReactDom.unmountComponentAtNode( document.getElementById( 'wpcom' ) );
-			renderLayout( context.store );
+			renderLayoutPromise( context.store ).then( () => {
```
Nope, I don't think so! Good call 😄

alexsanford commented on client/boot/project/wordpress-com.js:
```diff
@@ -247,12 +253,16 @@ export function setupMiddlewares( currentUser, reduxStore ) {
 			ReactDom.unmountComponentAtNode( document.getElementById( 'primary' ) );
 			ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
+			next();
+		} else {
```
Originally we had an `if ... else if` statement, and the call to `next()` was outside of the statement, so it would be run unconditionally after the block.

But now, in the `if` block, I need to call `next()` async, after the `Layout` is rendered. So I can't have a synchronous call to `next()` outside of the conditional statement. I needed to handle the `else` case, since we still need to call `next()` if we don't get into the `if` or the `else if` block.

nb replied:
👍

alexsanford commented on client/controller/index.web.js:
```diff
@@ -24,25 +23,64 @@ export { setSection, setUpLocale } from './shared.js';
+	return new Promise( ( resolve ) => {
+		if ( loggedIn ) {
+			// If user is logged in, async load the main layout
+			require.ensure( [], () => {
```
I tried using that first. The problem, as I understand it, is that many components expect to find the `#primary` or `#secondary` element on the page to render themselves into. So I needed to defer their rendering until those elements were available, and it's the `Layout` component that creates them.

I did this by defering the render through middleware. I don't think this would be possible with `AsyncLoad`. But I could be wrong! 😄

alexsanford replied:
@nb I think, however, the related [`asyncRequire`](https://github.com/Automattic/wp-calypso/blob/0ca65f4bddbefb8375f33fe7cbd5ddf736604597/server/bundler/babel/babel-plugin-transform-wpcalypso-async/test/index.js#L52) function would work. This would be beneficial since it falls back to a static `require` for a non-async environment. I'm going to try using that instead of `require.ensure`.

alexsanford replied:
@nb Done. I'm using `asyncRequire` now 👍

gziolo commented:
@alexsanford thanks for starting work on this one.

Wow, I didn't even expect that Layout contributes to over 20% of the build size! It's a huge impact. Have you evaluated if all section chunks remain about the same? It's highly possible that when you remove such big amount of code, every chunk now need to add shared dependencies that are now missing. It would be a really great idea to share all chunk sizes before and after this change to have an idea how impacts the rest of the app. I think it's acceptable trade-off to have other chunks a bit bigger, but it must remain sane :)

> I am worried about the general usefulness of the change. Our main goals is to render something, like a decent the shell as quickly as possible.

@nb it is something we definitely something we want to have in Calypso. It would greatly improve experience for logged out users as @alexsanford mentioned here:

> Granted, it's definitely not a sweeping performance improvement, although I think it's a pretty good improvement for a first-time visitor. First-time visitors are generally going to be logged out, so they don't need to download this fairly large chunk of Layout code until they signup or login.

I think it is going to be even more important when we enable server-side rendering for Login page. It is possible to fill in the form before all JS code is loaded and processed by the browser! It happened a few time when we were testing this new page ... That's why (Amber team) we believe this is a great change that is going to make user experience a bit better. 150kb code to load less when user visits us for the first time can make a difference.

> Theoretically, we could also extract the LayoutLoggedOut component very easily. I didn't do so for a couple of reasons:
> 
> It contributes much less size to the bundle. Using the Bonsai analysis tool, Layout has a "Weighted Size" of 2.29 MB, whereas LayoutLoggedOut has a "Weighted Size" of only 10.98 KB.
> 
> In the interest of keeping logged-out pages fast, I felt that loading LayoutLoggedOut async might have too much overhead. It is probably faster if we leave it in build.js.

`layout/logged-out.jsx` is imported in 3 different places. Did you make sure you moved all those 3 occurrences to its own chunk when doing calculations? Just double checking, it is possible that it is quite small :)

gziolo commented:
Another approach to consider is using [Running multiple entry points](https://webpack.github.io/docs/code-splitting.html#running-multiple-entry-points). At the moment we have this:
```js
webpackConfig.entry.build = path.join( __dirname, 'client', 'boot', 'app' );
```

we might experiment to have:

```js
webpackConfig.entry.build-logged-in = path.join( __dirname, 'client', 'boot', 'app.logged-in' );
webpackConfig.entry.build-logged-out = path.join( __dirname, 'client', 'boot', 'app.logged-in' );
```

and run `CommonsChunkPlugin` on those two. Then we would always load `build.js` containing common code first, and then proper chunk depending on the user's session state.

@scruffian did we try something like this before? I remember you saying that it caused some issues in the signup flow. Documentation sounds like it makes lots of sense :)

gziolo commented:
> Our main goals is to render something, like a decent the shell as quickly as possible.

@nb yes, agreed on that one. This one is super important 👍

alexsanford commented:
@gziolo 

> Have you evaluated if all section chunks remain about the same?

I didn't look really closely, no. I will share what I have, but I'm running into JavaScript heap memory issues trying to regenerate the `stats.json` files. Once I get it sorted, I'll post links to the reports here.

> layout/logged-out.jsx is imported in 3 different places. Did you make sure you moved all those 3 occurrences to its own chunk when doing calculations? Just double checking, it is possible that it is quite small :)

I didn't actually try it! I used Bonsai to analyze what the difference would be if I removed it, and it looked like the impact would be very small. So based on that I didn't bother to track down where I would need to remove it from. Might be worth testing though.

Weighted size of Layout:

<img width="999" alt="2017-06-13_2019-layout" src="https://user-images.githubusercontent.com/842193/27108887-fdd80732-5075-11e7-8c41-176c54b796d4.png">

Weighted size of LayoutLoggedOut:

<img width="999" alt="2017-06-13_2019-layout-logged-out" src="https://user-images.githubusercontent.com/842193/27108896-0826c28c-5076-11e7-9b73-454915e788d1.png">

> Do we really need async load layouts? It seems like they are essential to display html code to the user thus we should make sure it is server to the user as soon as possible.

Hmmm... no I don't think there's any real reason why it would need to be async loaded. We're always going to need it when we're logged in, so I think sending it down right away is a much better idea. I think that will also greatly simplify this PR!

scruffian commented:
@gziolo: we used to have separate logged in and logged out entry points, but we removed them in https://github.com/Automattic/calypso-pre-oss/pull/7870

The main reason for dropping it was that the logged out build was actually larger than the logged in one. That was because we introduced code splitting to the logged in build but not the logged out one. Having two entry points increases complexity and the burden of maintenance. Can we just make the initial entry point really really simple?

alexsanford commented:
I've pushed a few commits. I've addressed the comments above. Also:

- I'm now loading the new Layout chunk immediately for logged-in users. Please take a look when you get a chance.

- I'm still having issues generating `stats.json`. I'll get the reports posted here as soon as I work through these issues.

- For now, I'm going to avoid creating multiple entry points.

- I've left the "sausage making" intact on the last few commits. I can squash them later.

cc @nb @gziolo

gziolo commented:
@scruffian Thanks for letting us know 👍 

> This also solves an issue where the server is incapable of determining whether to send the logged in/out build on /start in the development environment (because the wordpress_logged_in cookies are not sent to calypso.localhost).

 Yes, it needs to be fixed before we even try to have 2 entry points. I think what @alexsanford proposes here makes the most sense taking into account our previous experience.

> For now, I'm going to avoid creating multiple entry points.

Yes, same as above. It is the best way to move forward.

nb commented:
Just to confirm I am reading the code right (haven’t been able to test yet extensively, yet), is this the current state?

- for logged out users we’re rendering the layout async, but we’re not loading it async – it’s just split out from other scripts; how does that affect loading speed?
- for logged out users we’re doing everything async – both loading and rendering; how does this affect loading speed?
- we’re not changing the dependency tree of the layout.

alexsanford commented:
@gziolo I've updated the bundle reports. I added the links to the PR description above, and I'll paste them here, too:


- [Before extracting `Layout`](https://alexsanford.gitlab.io/calypso-performance/report-before-extract-layout-0316d6b7.html)

- [After extracting `Layout`](https://alexsanford.gitlab.io/calypso-performance/report-after-extract-layout-28922e7c.html)

It looks like a couple of the bundles do grow pretty substantially after extracting `Layout` ☹️  Here are some bad ones (size in KB, gzipped):

- `post-editor`: 374.95 -> 463.59
- `devdocs`: 272.1 -> 356.66
- `themes`: 47.95 -> 129.13
- `settings`: 77.84 -> 117.54
- `posts-pages`: 70.4 -> 113.52

And there may be others. I don't like the ones that have grown by almost 100KB 😬 

Given that the logged-in pages are still going to have to load `build` and `layout`, I'm wondering if the price is too high for the benefit that we get for logged-out. What do you think, @gziolo? Maybe if we could find and manage to extract the common dependencies from `post-editor` it would make it better?

cc @nb

alexsanford commented:
> Just to confirm I am reading the code right (haven’t been able to test yet extensively, yet), is this the current state?
>
> - for logged out users we’re rendering the layout async, but we’re not loading it async – it’s just split out from other scripts; how does that affect loading speed?
> - for logged out users we’re doing everything async – both loading and rendering; how does this affect loading speed?
> - we’re not changing the dependency tree of the layout.

@nb Points 1 and 3 are correct (I'm assuming for point 1 you meant to say "logged in"). Point 2 is almost correct, except that we actually not loading `Layout` at all. That's where the bigger performance benefit is: less code loaded when logged out.

As for the loading speed, it's really hard for me to benchmark it, since I can currently only run it in development mode. But my guess is:

- Logged-in users might have an initial performance hit, because they will be downloading more JS. `build` is smaller, which is good, but other bundles have grown (see https://github.com/Automattic/wp-calypso/pull/14988#issuecomment-309169051)

- Logged-out users should see a performance boost because they will download 160KB less JS code. I don't think the growth of the other bundles will affect logged-out users significantly (@gziolo, do you agree?)

- Logged-in users will see an improvement in some repeat visits, since `build` and `layout` will be cached separately in the browser cache.

gziolo commented:
I tried today to execute `npm run analyze-bundles` on `master` branch and it looks like it throws error:
> FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

We probably reached the point when Webpack 1 doesn't scale anymore. We should put this one on hold until migration to Webpack 2 is finished, otherwise it's hard to validate changes.

> And there may be others. I don't like the ones that have grown by almost 100KB 😬

I don't think it is something that brings benefits in the long run. I would rather look for other ways to make this file smaller. When you look at what gets into async layout bundle you can notice that `notifications` client contributes the most. There is also `immutable`, which is being removed in other PRs. I think we can async load `notifications`.

alexsanford commented:
> I tried today to execute npm run analyze-bundles on master branch and it looks like it throws error:
>
> > FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory

@gziolo I hit the same error, and found a workaround. Please see [this post](https://calypsop2.wordpress.com/2017/06/16/webpack-stats-and-memory-usage/) 😄

gziolo commented:
You can now rebase with master. @DanReyLop merged code that fixes `npm run analyze-bundles`.

I opened alternative patch which tries to lazy load components that get bundled into `async-load-layout`. I picked them based on the output you provided in https://alexsanford.gitlab.io/calypso-performance/report-after-extract-layout-28922e7c.html.

alexsanford commented:
> I opened alternative patch which tries to lazy load components that get bundled into async-load-layout. I picked them based on the output you provided in https://alexsanford.gitlab.io/calypso-performance/report-after-extract-layout-28922e7c.html.

@gziolo I created some new reports. One is for your branch with the dependencies extracted, and for the other one I applied my changes on top of your branch. Here are the reports:

- [With dependencies extracted](https://alexsanford.gitlab.io/calypso-performance/report-extract-deps-no-layout-d38d08a5.html)

- [With dependencies and full Layout extracted](https://alexsanford.gitlab.io/calypso-performance/report-extract-deps-and-layout-20654260.html)

Pulling out the Layout still reduces the size of the build by about 100KB gzipped. However, it increases the size of `post-editor` and `devdocs` by almost 100KB... 🤷‍♂️

stephanethomas commented:
@alexsanford: what's the exact status of this pull request?

Amber needs to be able to customize the [Calypso-native `Login` page](https://wordpress.com/log-in) to look **exactly** like the [WooCommerce `Login` page](https://public-api.wordpress.com/oauth2/authorize?response_type=code&client_id=50916&redirect_uri=https%3A%2F%2Fwoocommerce.com%2Fwc-api%2Fwpcom-signin%3Fnext%3Dmyaccount&blog_id=0&wpcom_connect=1). This means we must be able to use a layout that is different from the default one, without the usual masterbar and footer. As such, this pull request seems like a pre-requisite.

alexsanford commented:
@stephanethomas I put together this PR as part of my trial, but ended up moving in a different direction. I will close this one.

You might want to check out #15287 though. It pulls out some of the larger pieces a little less intrusively, so it may or may not affect what you're doing.

cc @nb
